website/blog/2023-04-04-presto-hudi-deep-dive.md

```diff
@@ -0,0 +1,110 @@
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/get-splits-call.png)
+*Figure 1: listStatus() gets called internally from getSplits().*
+
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/make-split-call.png)
+*Figure 2: makeSplit() call is highlighted as part of getSplits() implementation.*
+
+The above snippets are taken from FileInputFormat class. Please note that the lines are highlighted where these calls are being made. 
+
+## Hudi table types
+
+Hudi supports 2 types of tables out of the box, namely, Copy on Write (CoW) and Merge on Read (MoR). CoW is the simpler of the two in terms of the reading and writing complexities involved. Upserts made to this table type result in newer versions of columnar parquet files which can be read using the ParquetPageSource used internally by Presto.
+MoR involves writing incoming upserts to delta log files in avro format by default. These log files are later compacted to create new base parquet files. Since this involves row and columnar format files, this table type supports Read Optimized (RO) and Real Time (RT) queries.
+
+To be able to query these tables, the table metadata needs to be synced with Hive or Glue metastore. Syncing MoR tables with the configured metastore results in the creation of 2 table types, `<table_name>_ro` and `<table_name>_rt`. They are internally differentiated by setting the config `hoodie.query.as.ro.table`. As the name suggests, the former is used for read optimized queries involving only base parquet files, while the latter is used for real time queries involving the merging of delta log files and base parquet files on the fly.
+
+From Presto’s point of view, CoW table and `_ro` version of a MoR table behave the same way. However `_rt` version of MoR table type requires additional work as described in the next section.
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/hudi-mor-ro-1.png)
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/hudi-mor-ro-2.png)
+*Figure 3: Snapshot showing details about _ro table.*
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/hudi-mor-rt-1.png)
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/hudi-mor-rt-2.png)
+*Figure 4: Snapshot showing details about _rt table.*
+
+The above figures show the difference between the 2 versions of MoR tables in terms of table properties and input formats involved.
+
+As is clearly visible, Hudi syncs the CoW and `_ro` tables with `HoodieParquetInputFormat` while `_rt` tables are registered with `HoodieParquetRealtimeInputFormat`. 
+
+### Merge on Read table type
+
+Currently presto supports read optimized and real time queries for MoR tables via the hive connector. After the parser, planner and optimizer phases, splits are generated by Presto which are used for creating presto pages to be rendered back to the client. We focus on this split generation phase and the record reader involved thereafter in the next sections. 
```

Comment:
```suggestion
Currently presto supports read optimized and real time queries for MoR tables via the hive as well as the hudi connector. After the parser, planner and optimizer phases, splits are generated by Presto which are used for creating presto pages to be rendered back to the client. We focus on this split generation phase and the record reader involved thereafter in the next sections. 
```

Reply:
Do you want me to do this change? At other places in this blog, I am only talking about hudi support via hive connector. I feel adding it explicitly for MoR tables might result in misunderstandings that hudi connector is lacking in other features that we are mentioning, for example, CoW reading etc. Happy to hear your thoughts on this @codope

Reply:
Makes sense. Let's keep it the way you have written but let's also make it clear in the title as suggested above. We can even remove `via the hive connector` part.

Reply:
sure, that should make it more clear